servers/src/main/java/tachyon/worker/block/evictor/LRUEvictor.java

```diff
@@ -84,10 +95,58 @@ public EvictionPlan freeSpace(long availableBytes, BlockStoreLocation location)
 
     // enough free space
     if (dirCandidates.maxAvailableBytes() >= toEvictBytes) {
+      // candidate blockIds for eviction from current tier, sorted from less recently used to more
+      // recently used
       toEvict = dirCandidates.toEvict();
+      // reverse list so that the more recently used blocks are moved to next tier first
+      Collections.reverse(toEvict);
+      // move as many blocks to next tier as possible
+      int nextTierLevel = mMeta.getBlockMeta(toEvict.get(0)).getBlockLocation().tierLevel() + 1;
+      List<StorageTier> tiers = mMeta.getTiers();
+      int nTotalTiers = tiers.size();
+      while (nextTierLevel < nTotalTiers && toEvict.size() > 0) {
+        // find space in next tier
+        List<StorageDir> dirs = tiers.get(nextTierLevel).getStorageDirs();
+        for (StorageDir dir : dirs) {
+          BlockStoreLocation dest = dir.toBlockStoreLocation();
+          Iterator<Long> blocks = toEvict.iterator();
+          long remainBytes = dir.getAvailableBytes();
+          while (blocks.hasNext() && remainBytes >= 0) {
+            long blockId = blocks.next();
+            long blockSize = mMeta.getBlockMeta(blockId).getBlockSize();
+            if (blockSize <= remainBytes) {
+              // the block can be moved to the dir
+              toMove.add(new Pair<Long, BlockStoreLocation>(blockId, dest));
+              blocks.remove();
+              remainBytes -= blockSize;
+            }
+          }
+        }
+        nextTierLevel ++;
+      }
+
       for (Long blockId : toEvict) {
         mLRUCache.remove(blockId);
       }
+
+      // assure all blocks are in the store, if not, remove both from plan and lru cache
```

Comment:
If the block is removed from the plan, will the plan still satisfy the space requirement?